Adds support for geo-bounds filtering in geogrid aggregations

talevy opened the pull request:
It is fairly common to filter the geo point candidates in
geohash_grid and geotile_grid aggregations according to some
viewable bounding box. This change introduces the option of
specifying this filter directly in the tiling aggregation.

This is even more relevant to `geo_shape` where the bounds will restrict
the shape to be within the bounds

this optional `bounds` parameter is parsed in an equivalent fashion to 
the bounds specified in the geo_bounding_box query.

example:

```
POST /museums/_search?size=0
{
    "aggregations" : {
        "tiles-in-bounds" : {
            "geohash_grid" : {
                "field" : "location",
                "precision" : 8,
                "bounds": {
                  "top_left" : "53.4375, 4.21875",
                  "bottom_right" : "52.03125, 5.625"
                }
            }
        }
    }
}
```

talevy committed:
Adds support for geo-bounds filtering in geogrid aggregations

It is fairly common to filter the geo point candidates in
geohash_grid and geotile_grid aggregations according to some
viewable bounding box. This change introduces the option of
specifying this filter directly in the tiling aggregation.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Geo)

talevy commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/CellIdSource.java:
```diff
@@ -81,20 +87,32 @@ public SortedBinaryDocValues bytesValues(LeafReaderContext ctx) {
+                    if (Double.isNaN(topLeft.lon()) ||
+                            (target.lon() < bottomRight.lon() && target.lon() > topLeft.lon()
+                            && target.lat() < bottomRight.lat() && target.lat() > topLeft.lat())) {
+                        values[j++] = encoder.encode(target.getLon(), target.getLat(), precision);
```
hey @iverase. I thought it a good idea to introduce this `bounds` idea directly to geotile_grid in master first.
The check has some work to do with regards to normalization and dateline handling I believe. I was wondering if there are any geo-utils already around for checking if a point is in a bounding box that I should be re-using here.

iverase replied:
Nothing that is worthy.

talevy replied:
OK. I wrote something and added it to GeoUtils

talevy commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/CellIdSource.java:
```diff
@@ -81,20 +87,32 @@ public SortedBinaryDocValues bytesValues(LeafReaderContext ctx) {
+                int j = 0;
                 for (int i = 0; i < docValueCount(); ++i) {
                     org.elasticsearch.common.geo.GeoPoint target = geoValues.nextValue();
-                    values[i] = encoder.encode(target.getLon(), target.getLat(), precision);
```
another question I have @iverase is how it might be best to incorporate these bounds during the rasterization of shapes (for the geoshape-doc-values branch). It seems like there can be a few different situations that should be concerned differently. I am thinking it makes sense to pass the bounds to the GeoGridTiler and use it to clip the tile the tile-queries such that tiles that intersect the bounds are made smaller to their overlapping shape. tile-queries that are disjoint from the bounds are immediately ignored and no `relate` call is made. If this works, I think it would be nice since then there are no additional calls to relate in any pre-processing way and it does not complicate the tiling logic too much. What do you think?

iverase replied:
We will have to check performance but my first the on this would be first to check if the shape intersects the provided bounding box. If true then go to the tiler and again filter out the tiles that do not intersects the provided bounds.

iverase replied:
Actually we can skip the first?

talevy replied:
I will check, but even if we do the initial intersection check, we need to clip each tile according to the provided bounds. That is why I thought the initial check would not be too helpful most times that the bounds provided to the geotile_grid match that which is passed to the geo_bounding_box filter.

nknize commented on docs/reference/aggregations/bucket/geohashgrid-aggregation.asciidoc:
```diff
@@ -191,6 +191,62 @@ var bbox = geohash.decode_bbox('u17');
+                "field" : "location",
+                "precision" : 8,
+                "bounds": {
+                  "top_left" : "u17",
```
Having the same geohash for `top_left` and `bottom_right` is a bit confusing from an API standpoint. I think if a user explicitly defines `top_left` and `bottom_right` we should expect lat,lon decimal degrees. For example:
```
"bounds" : {  
    "top_left" : "52.71, 4.89", 
    "bottom_right" : "52.69, 4.91" 
} 
```
If the user defines by way of a geohash it should be something simple like `"bounds"` defined as a string instead of an object. For example, `"bounds" : "u17"` and then we compute the spatial bounds of the geohash cell.

talevy replied:
I just figured I would use this as the example. both are supported. I'll update to use the lat/lon

nknize commented on docs/reference/aggregations/bucket/geotilegrid-aggregation.asciidoc:
```diff
@@ -162,6 +162,62 @@ POST /museums/_search?size=0
+                "precision" : 22,
+                "bounds": {
+                  "top_left" : "52.4, 4.9",
+                  "bottom_right" : "52.3, 5.0"
```
:+1:

nknize commented on server/src/main/java/org/elasticsearch/common/geo/GeoUtils.java:
```diff
@@ -93,6 +93,30 @@ public static boolean isValidLongitude(double longitude) {
+     * @param right the eastern longitude of the bounding box
+     * @return whether the point (lon, lat) is in the specified bounding box
+     */
+    public static boolean pointInBounds(double lon, double lat, double top, double bottom, double left, double right) {
```
Can we use Lucene's [Rectangle#containsPoint](https://github.com/apache/lucene-solr/blob/2feeb88c294030be654183d22a8e503bbbf8b293/lucene/core/src/java/org/apache/lucene/geo/Rectangle.java#L88) static method instead and remove this?

talevy replied:
it didn't look like Rectangle#containsPoint works for Rectangles that cross the dateline? or I might be mis-using it, but I would expect this test to pass

```
    public void testDatelineCrossingRectangleContainsPoint() {
        assertTrue(Rectangle.containsPoint(67.35, 178.41, 175.07, -174.95, 68.57, 66.19));
    }
```

nknize commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/GeoTileGridValuesSourceBuilder.java:
```diff
@@ -61,13 +70,27 @@ static GeoTileGridValuesSourceBuilder parse(String name, XContentParser parser)
     GeoTileGridValuesSourceBuilder(StreamInput in) throws IOException {
         super(in);
         this.precision = in.readInt();
+        if (in.getVersion().onOrAfter(Version.V_8_0_0)) {
```
should we backport to 7.x?

talevy replied:
for sure. I tend to go this strategy when doing bwc changes

https://github.com/elastic/elasticsearch-team/pull/318/files#diff-60ab984f14487976eab0e31f39880a3bR30

talevy replied:
but this does remind me, I do not have enough tests for this bwc change. so I will get on that

nknize commented on server/src/test/java/org/elasticsearch/index/search/geo/GeoUtilsTests.java:
```diff
@@ -692,6 +694,34 @@ public void testParseGeoPointGeohashPositions() throws IOException {
             GeoUtils.EffectivePoint.BOTTOM_LEFT), new GeoPoint(42.890625, -71.3671875));
     }
 
+    public void testPointInBounds() {
```
I think this goes away if we use lucene's `Rectangle#containsPoint` method since it''s unit tested by lucene?

talevy replied:
I do not see any explicit tests for this method in Lucene

https://github.com/apache/lucene-solr/search?q=containsPoint&unscoped_q=containsPoint

are there indirect tests for this?

nknize commented on server/src/test/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridAggregatorTestCase.java:
```diff
@@ -112,17 +122,33 @@ public void testWithSeveralDocs() throws IOException {
 
-    private void testCase(Query query, String field, int precision, CheckedConsumer<RandomIndexWriter, IOException> buildIndex,
-                          Consumer<InternalGeoGrid<T>> verify) throws IOException {
+    public void testBounds() throws IOException {
```
Can we randomize the test? e.g., `GeometryTestUtils.randomRectangle` and add doc value points somewhere near the middle of the random generated rectangle?

talevy replied:
for sure. I'll randomize. I felt like I could by-pass randomization here due to 
the underlying `pointInBounds` already having randomized tests in `testPointInBounds`

nknize reviewed (changes requested):
Overall it looks good. Left a few minor thoughts and suggestions. At minimum I think we should use the Lucene test utilities where we can. Separate from that I think we can introduce a little more randomization to the tests.

talevy committed:
Centralize BoundingBox logic to a dedicated class

Both geo_bounding_box query and geo_bounds aggregation have
a very similar definition of a "bounding box". A lot of this
logic (serialization, xcontent-parsing, etc) can be centralized
instead of having separated efforts to do the same things

talevy committed:
add tests

talevy committed:
fix checkstyle

talevy committed:
rename BoundingBox -> GeoBoundingBox

elasticmachine committed:
Merge branch 'master' into reuse-bounding-box

talevy committed:
update to geo-bounding-box and modify tests

talevy committed:
Merge remote-tracking branch 'elastic/master' into geogrid-bounds

talevy committed:
fix unused

talevy committed:
fix pointInBounds tests

talevy committed:
Merge remote-tracking branch 'elastic/master' into geogrid-bounds

talevy committed:
add bwc serialization tests

iverase reviewed (commented):
I think this is almost there, just want to think if we should use different implementations for the bounded and unbounded case when iterating doc values.

talevy committed:
Merge remote-tracking branch 'elastic/master' into geogrid-bounds

talevy committed:
split out bounded and unbounded cellvalues

talevy committed:
fix forgotten bug

talevy commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/CellIdSource.java:
```diff
@@ -55,7 +58,7 @@ public boolean isFloatingPoint() {
     @Override
     public SortedNumericDocValues longValues(LeafReaderContext ctx) {
-        return new CellValues(valuesSource.geoPointValues(ctx), precision, encoder);
+        return new CellValues(valuesSource.geoPointValues(ctx), precision, geoBoundingBox, encoder);
```
I am not sure how much of a hit calling `isUnbounded` or even checking if the object is null would be. I also see this check as unnecessary to be done for each value, when it is known up front. I've split out the logic to Bounded and Unbounded cell values.

talevy committed:
fix other bug

iverase reviewed (approved):
lgtm

talevy commented:
@elasticmachine update branch

thanks Ignacio!

talevy commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into geogrid-bounds
